pandas/io/parsers.py

```diff
@@ -2123,9 +2123,27 @@ def _clean_mapping(mapping):
         else:
             clean_dtypes = _clean_mapping(self.dtype)
 
-        return self._convert_to_ndarrays(data, self.na_values, self.na_fvalues,
-                                         self.verbose, clean_conv,
-                                         clean_dtypes)
+        # Apply NA values.
+        clean_na_values = {}
+        clean_na_fvalues = {}
+
+        if isinstance(self.na_values, dict):
```

Comment:
Hmmm...perhaps, but the logic is very coupled with the class attributes of the C and Python parsers.  In addition, the way the logic is applied (in bulk for Python, per iteration in C), merging the two would not be straightforward.